libbeat/processors/add_host_metadata/add_host_metadata.go

```diff
@@ -63,39 +63,42 @@ func newHostMetadataProcessor(cfg *common.Config) (processors.Processor, error)
 	p := &addHostMetadata{
 		info:   h.Info(),
 		config: config,
+		data:   common.NewMapStrPointer(nil),
 	}
 	return p, nil
 }
 
 // Run enriches the given event with the host meta data
 func (p *addHostMetadata) Run(event *beat.Event) (*beat.Event, error) {
 	p.loadData()
-	event.Fields.DeepUpdate(p.data.Clone())
+	event.Fields.DeepUpdate(p.data.Get())
```

Comment:
This introduces another potential race. Better: `p.data.Get().Clone()`

Reply:
just for my understanding, does that mean that the write on data could be coming from some processor in the pipeline? this processor itself is only using set, while treating the inserted MapStr as immutable.

Reply:
Right. Processors or even the output can add/modify the event. E.g. users adding more fields to host (this did actually happen):

```
fields:
  host.whatever: x
fields_under_root: true
```

If an user does this to one input only, then the shared data must not be updated at all (so to guarantee data consistency of other modules/inputs). Plus: modifying shared data can race with the serialisation in our outputs.

Working with events and shared data + custom processors is full of pitfalls and potential race conditions. We will have to think about some 'event-type' that can share data (such that we do not create this much garbage), but yet, does prevent developers from accidentally overwriting shared data.

As processors can be run in parallel, caching with concurrent updates is another pitfall in processors.

These issues are known to the Core Team. We hope to provide API/libs dealing with these potential races/pitfalls for you, such that input/processor developers don't have to deal with these kind of issues. At times It's hard to get this right, even if you are aware of these issues.